### PR TITLE
Regularization

### PR DESCRIPTION
**Pull request by blengerich**
Allow for regularization on sample models by passing in a lambda function to a `ContextualizedRegressionBase` constructor as arg `sample_model_regularizer`. Default is `zero_vector`, defined in functions.py.

### CODE REVIEW

File: contextualized/functions.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import torch
 import torch.nn.functional as F
 
-
+zero_vector = lambda x, *args: torch.zeros((len(x), 1))
```
Comment (cnellington) [inline contextualized/functions.py:4]: Names should relate to regularization losses, maybe `no_reg`? 
Also compile link functions and regularization losses into lists `LINK_FUNCTIONS` AND `REGULARIZERS` for clarity

File: contextualized/regression.py [hunk 2]
```diff
@@ -358,10 +358,13 @@ def forward(self, C, T):
...
 class ContextualizedRegressionBase(pl.LightningModule):
-    def __init__(self, *args, learning_rate=1e-3, link_fn=linear_fn(), **kwargs):
+    def __init__(self, learning_rate=1e-3, link_fn=linear_fn(), loss_fn=MSE,
+                 sample_model_regularizer=zero_vector, *args, **kwargs):
         super().__init__()
-        self.link_fn = link_fn
         self.learning_rate = learning_rate
+        self.link_fn = link_fn
+        self.loss_fn = loss_fn
+        self.sample_model_regularizer = sample_model_regularizer
```
Comment (cnellington) [inline contextualized/regression.py:367]: Should be `model_regularizer` based on the current convention of metamodels and models.

File: contextualized/regression.py [hunk 3]
```diff
@@ -416,6 +419,9 @@ def test_step(self, batch, batch_idx):
         self.log_dict({'test_loss': loss})
         return loss
 
+    def predict_from_sample_models(self, X, beta_hat, mu_hat):
```
Comment (cnellington) [inline contextualized/regression.py:422]: `predict_from_models`

**Review by cnellington — CHANGES_REQUESTED** [review]

**Comment by cnellington** [comment]
Linking to #60

File: contextualized/regression.py (hunk 3 above)
Comment (blengerich) [reply contextualized/regression.py:422]: Why private? I can imagine post-hoc analyses that may want to use this function (e.g. sensitivity anlalyses) and it's not clear to me why it should be private.

File: contextualized/functions.py (hunk 1 above)
Comment (blengerich) [reply contextualized/functions.py:4]: I don't have strong feelings either way; putting the organization in `__init__.py` exports and highlights the available utilities.

**Pull request merged**